### DISCUSSION TITLE
Tabs: Persist currently selected tab in URL

### DISCUSSION THREAD

**Discussion by isarphilipp** [livewire/flux#314] [Feature Requests] [answered]
Introduce a new option `<flux:tabs to-url="my-key">` that pushes the name of the currently selected tab as query param to the URL.
Clicking a tab with the `name="general"` would result in this URL: `https://flux.test/page-with-tabs?my-key=general`

It would be a nice convenience feature when tabs are used for subnavigation.

**Comment by lucasromanojf** [comment]
You can use `wire:model` on your `flux:tabs` component and configure your binded Livewire property to append to URL

**Reply by benjibee** [reply]
Caleb gave a good example of this in his Flux introduction over on Vimeo. [Here's a link](https://vimeo.com/1011642608/478575dea1?share=copy#t=845.415) to the part of the video where he walks you through it.

**Reply by RauchenwaldC** [reply]
Just out of curiosity, is there any way to set the current active tab based on a GET parameter without using <flux:tab.group> in a Livewire component?

**Reply by jeffchown** [reply]
@RauchenwaldC Not sure what you're asking...  
Can you provide a snippet, with an explanation of what you're trying to accomplish, for clarification?

**Reply by RauchenwaldC** [reply]
@jeffchown 
```
<flux:tab.group>
    <flux:tabs wire:model="tab">
        <flux:tab name="profile">Profile</flux:tab>
        <flux:tab name="account">Account</flux:tab>
        <flux:tab name="billing">Billing</flux:tab>
    </flux:tabs>

    <flux:tab.panel name="profile">...</flux:tab.panel>
    <flux:tab.panel name="account">...</flux:tab.panel>
    <flux:tab.panel name="billing">...</flux:tab.panel>
</flux:tab.group>
```

I'd like to be able to define which of the three tabs is shown based on a GET parameter like `?tab=account` when I'm using this in a regular old-fashioned Blade template and not a Livewire component.

All my attempts of doing something like:

`<flux:tab name="account" {{ request('tab') === 'account' ? 'data-selected="true"' :: '' }}>Account</flux:tab>`

result in things like:

`syntax error, unexpected double-quoted string ""`

or `<flux:tab name="account" @if(request('tab')==='account' ) data-selected="true" @endif>Account</flux:tab>`
results in `syntax error, unexpected token "endif", expecting end of file`

I'm not sure if I'm missing something here. I would've thought that Tabs also support `:current="request()->is('/')"` like, for example, the Flux Navbar element.

**Reply by PrinceMinky** [reply]
In your component add the URL attribute above your public tab variable. 

```

use Livewire\Attributes\Url;

#[Url]
public $tab = '';
```

Also ensure you add `wire:model="tab"` to the flux ui tab component.

**Reply by RauchenwaldC** [reply]
> I'd like to be able to define which of the three tabs is shown based on a GET parameter like ?tab=account when I'm using this in a regular old-fashioned Blade template and **not a Livewire component**.

**Reply by jeffchown** [reply]
@RauchenwaldC Try this:

```
<flux:tab.group>
    <flux:tabs>
        <flux:tab name="profile" :selected="request('tab') == 'profile'">Profile</flux:tab>
        <flux:tab name="account" :selected="request('tab') == 'account'">Account</flux:tab>
        <flux:tab name="billing" :selected="request('tab') == 'billing'">Billing</flux:tab>
    </flux:tabs>

    <flux:tab.panel name="profile">...</flux:tab.panel>
    <flux:tab.panel name="account">...</flux:tab.panel>
    <flux:tab.panel name="billing">...</flux:tab.panel>
</flux:tab.group>
```

**Reply by PrinceMinky** [reply]
> > I'd like to be able to define which of the three tabs is shown based on a GET parameter like ?tab=account when I'm using this in a regular old-fashioned Blade template and **not a Livewire component**.

Flux is a Livewire UI component. 
"The official Livewire UI component library"

Can't say I've ever tried to use Flux without Livewire.

**Reply by PrinceMinky** [reply]
I tried to recreate what I thought you was asking for.

I think you're wanting the URL to update as you go through each tab, which, in turn allows you to navigate using the back page functionality built into the browser - as well as allowing users to click a link elsewhere and navigate to the page with the correct tab applied.

I created this thinking that this would work: _(which it didn't)_
```
            <flux:tab.group>
                <flux:tabs>
                    <flux:tab name="profile" :href="url()->current() . '?tab=profile'" :selected="request('tab') == 'profile'">Profile</flux:tab>
                    <flux:tab name="account" :href="url()->current() . '?tab=account'" :selected="request('tab') == 'account'">Account</flux:tab>
                    <flux:tab name="billing" :href="url()->current() . '?tab=billing'" :selected="request('tab') == 'billing'">Billing</flux:tab>
                </flux:tabs>
            
                <flux:tab.panel name="profile">...</flux:tab.panel>
                <flux:tab.panel name="account">...</flux:tab.panel>
                <flux:tab.panel name="billing">...</flux:tab.panel>
            </flux:tab.group>
```

So I tried this: _(navigating to the page with the correct tab applied on load works)_
```
            <flux:tab.group>
                <flux:tabs>
                    <flux:tab name="profile" :selected="request('tab') == 'profile'">Profile</flux:tab>
                    <flux:tab name="account" :selected="request('tab') == 'account'">Account</flux:tab>
                    <flux:tab name="billing" :selected="request('tab') == 'billing'">Billing</flux:tab>
                </flux:tabs>
            
                <flux:tab.panel name="profile">...</flux:tab.panel>
                <flux:tab.panel name="account">...</flux:tab.panel>
                <flux:tab.panel name="billing">...</flux:tab.panel>
            </flux:tab.group>
```

I couldn't get it to work when using `href` as it changed the tab from a button to an anchor tag.

**Reply by jeffchown** [reply, marked as answer]
If you want to not only set the current tab from the url's query string on load, but update the query string when a tab is clicked, you can use:

```blade
<div x-data="{
    setTab(tabName) {
        const url = new URL(location);

        url.searchParams.set('tab', tabName);
        history.pushState({}, '', url);
    },
}">
    <flux:tab.group>
        <flux:tabs>
            <flux:tab name="profile" x-on:click="setTab('profile')" :selected="request('tab') == 'profile'">Profile</flux:tab>
            <flux:tab name="account" x-on:click="setTab('account')" :selected="request('tab') == 'account'">Account</flux:tab>
            <flux:tab name="billing" x-on:click="setTab('billing')" :selected="request('tab') == 'billing'">Billing</flux:tab>
        </flux:tabs>

        <flux:tab.panel name="profile">...</flux:tab.panel>
        <flux:tab.panel name="account">...</flux:tab.panel>
        <flux:tab.panel name="billing">...</flux:tab.panel>
    </flux:tab.group>
</div>
```

**Reply by RauchenwaldC** [reply]
@jeffchown Thank you so much.

Might be nice to add the information about `:selected` to the docs as well. Now, in hindsight, it makes sense that it's the `selected` attribute, but since I didn't notice the `data-selected` that's mentioned at the bottom of the docs I decided to try with `current` since that's what's used in other elements.

**Reply by jeffchown** [reply]
You're welcome, @RauchenwaldC !  Yes, some mention of the `:selected` attribute in the docs could be helpful.  In the case of `flux:tabs` its usage is a bit of an edge case (as most people will `wire:model` a property to accomplish the same result), but for those not using Livewire, it is (obviously ;)) useful.

If this solves your issue/question, please mark the discussion as answered so everyone knows it's resolved.  Thanks.

**Reply by RauchenwaldC** [reply]
> If this solves your issue/question, please mark the discussion as answered so everyone knows it's resolved. Thanks.

Wish I could, but the discussion was started by @isarphilipp

**Reply by jeffchown** [reply]
Ha!  I just realized that! 🙃

**Reply by jeffchown** [reply]
@isarphilipp Let us know if https://github.com/livewire/flux/discussions/314#discussioncomment-12823032 solves your initial question.
